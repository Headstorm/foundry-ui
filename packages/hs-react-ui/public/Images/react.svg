<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 19.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_2" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 width="561.8px" height="499.9px" viewBox="-80.4 151.2 561.8 499.9" style="enable-background:new -80.4 151.2 561.8 499.9;"
	 xml:space="preserve">
<style type="text/css">
	.st0{fill:#939598;}
</style>
<circle class="st0" cx="200.5" cy="400.6" r="50.2"/>
<path class="st0" d="M200.5,515.6c-70.5,0-132.1-8.3-178.2-24.1c-29.9-10.2-55.3-23.8-73.4-39.3c-19.2-16.4-29.4-34.3-29.4-51.6
	c0-33.2,36.4-65.7,97.5-86.9c50-17.4,115.2-27.1,183.4-27.1c67,0,131.3,9.4,181,26.6c29.1,10,53.6,23,71,37.4
	c18.9,15.8,28.9,33.1,28.9,50c0,34.5-40.7,69.4-106.3,91.1C328.6,507.1,266.6,515.6,200.5,515.6z M200.5,310.6
	c-64.7,0-128.7,9.4-175.5,25.7c-40.5,14.1-81.4,39.5-81.4,64.3c0,10,7.4,21.8,21,33.4c15.8,13.5,38.4,25.6,65.5,34.8
	c43.6,14.9,102.6,22.8,170.4,22.8c63.6,0,122.9-8,167-22.7c61.7-20.5,89.9-49.8,89.9-68.3c0-9.5-7.2-20.7-20.3-31.6
	c-15.1-12.6-37.1-24.1-63.4-33.2C326.4,319.6,264.9,310.6,200.5,310.6z"/>
<path class="st0" d="M86.6,650.8c-10.2,0-19.2-2.2-26.8-6.6C31,627.6,21.1,579.8,33.1,516.3c9.9-52.1,34.1-113.3,68.2-172.4h0
	c33.5-58,73.7-109,113.4-143.5c23.2-20.2,46.7-35,67.9-42.8c23.1-8.5,43.1-8.5,57.7-0.1c29.9,17.2,39.8,70,25.8,137.6
	c-9.9,48-33.5,105.9-66.5,163.2c-35.2,61-73.2,110.2-109.9,142.3c-23.8,20.8-48.3,36-70.7,43.9C107.4,648.7,96.4,650.8,86.6,650.8z
	 M122.1,355.9c-32.3,56-56.2,116.1-65.4,164.9c-8,42.1-6.5,90.2,15,102.6c16.1,9.3,55,0.5,102.3-40.9c34.7-30.3,71-77.5,104.9-136.2
	c31.8-55.1,54.4-110.5,63.8-156c6-29.1,7.3-55.5,3.8-76.4c-3-17.9-9.4-30.6-18.2-35.6c-8.2-4.7-21.5-4.1-37.5,1.8
	c-18.5,6.8-39.4,20.1-60.4,38.4C192.8,251.3,154.3,300.1,122.1,355.9L122.1,355.9z"/>
<path class="st0" d="M314.4,651.1c-27.2,0-61.7-16.4-97.7-47.4c-40.2-34.6-81.1-86.1-115.3-145.2v0c-33.6-58-57.6-118.3-67.7-170
	c-5.9-30.2-7-57.9-3.2-80.2c4.2-24.3,14.1-41.6,28.8-50.1c29.8-17.3,80.5,0.5,132.1,46.4c36.6,32.5,75,81.9,108.1,139.1
	c35.3,61,59,118.5,68.4,166.3c6.1,31,7.1,59.8,2.8,83.2c-4.6,24.9-15,42.6-30,51.3C333.2,648.9,324.3,651.1,314.4,651.1z
	 M122.2,446.5c32.4,56,72.6,106.7,110.2,139c32.5,28,74.9,50.6,96.4,38.2c16.1-9.3,27.9-47.4,15.7-109
	c-9-45.2-31.7-100.2-65.6-158.9c-31.9-55.1-68.6-102.4-103.3-133.2c-22.2-19.7-44.4-34.1-64.2-41.5c-17-6.4-31.2-7.1-39.9-2.1
	c-8.2,4.7-14.3,16.6-17.2,33.4c-3.3,19.4-2.3,44.2,3.1,71.5C66.9,333,89.9,390.7,122.2,446.5L122.2,446.5z"/>
</svg>
