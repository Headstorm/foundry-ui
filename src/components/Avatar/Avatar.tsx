/* eslint-disable @typescript-eslint/no-non-null-assertion */
import styled from 'styled-components';
import React, { ReactNode } from 'react';
import { darken, lighten, readableColor } from 'polished';
import { StyledBaseSpan } from '../../htmlElements';
import { SubcomponentPropsType, StyledSubcomponentType } from '../commonTypes';
import Skeleton from '../Skeleton/Skeleton';
import { useTheme } from '../../context';

export type AvatarContainerProps = {
  size: number;
  borderRadiusPercent: number;
  imgURL?: string;
  color: string;
  isLoading: boolean;
};

export type AvatarTextProps = {
  size: number;
  textColor: string;
};

export type AvatarProps = {
  placeholder?: string | ReactNode;
  children?: ReactNode;
  imgURL?: string;
  size?: number;
  color?: string;
  borderRadiusPercent?: number;
  isLoading?: boolean;

  StyledContainer?: StyledSubcomponentType;
  StyledText?: StyledSubcomponentType;
  StyledShimmer?: StyledSubcomponentType;
  containerProps?: SubcomponentPropsType;
  textProps?: SubcomponentPropsType;
  shimmerProps?: SubcomponentPropsType;
  containerRef?: React.RefObject<HTMLDivElement>;
  textRef?: React.RefObject<HTMLSpanElement>;
  shimmerRef?: React.RefObject<HTMLDivElement>;
};

export const AvatarContainer = styled(Skeleton.Container)`
  ${({ size, borderRadiusPercent, imgURL, isLoading, color }: AvatarContainerProps) => {
    if (imgURL && !isLoading) {
      return `
        border-radius: ${borderRadiusPercent}%;
        overflow: hidden;
        width: ${size}em;
        height: ${size}em;
        background-image: url(${imgURL});
        background-size: cover;
      `;
    }
    return `
      display: flex;
      border-radius: ${borderRadiusPercent}%;
      padding: 1em;
      width: ${size}em;
      height: ${size}em;
      background-color: ${color};
      justify-content: center;
      align-items: center;
    `;
  }}
`;

export const AvatarText = styled(StyledBaseSpan)`
  ${({ size, textColor }: AvatarTextProps) => {
    return `
      color: ${textColor};
      font-size: ${size / 3}em; 
      font-weight: 600;
      `;
  }};
`;

export const AvatarShimmer = styled(Skeleton.Shimmer)`
  ${({ borderRadiusPercent }: Pick<AvatarProps, 'borderRadiusPercent'>) =>
    `
      border-radius: ${borderRadiusPercent}%;
    `}
`;

const Avatar = ({
  placeholder,
  children,
  imgURL,
  size = 1.5,
  borderRadiusPercent = 50,
  color: initialColor,
  isLoading = false,
  StyledContainer = AvatarContainer,
  StyledText = AvatarText,
  StyledShimmer = AvatarShimmer,
  textProps = {},
  containerProps = {},
  shimmerProps = {},
  containerRef,
  textRef,
  shimmerRef,
}: AvatarProps): JSX.Element => {
  const { colors } = useTheme();
  const color = initialColor || colors.grayXlight;
  const fontColor = readableColor(color, colors.grayMedium, colors.background);
  const shimmerColor = fontColor === colors.background ? lighten(0.2, color!) : darken(0.2, color!);
  return (
    <Skeleton
      isLoading={isLoading}
      color={shimmerColor}
      StyledContainer={StyledContainer}
      containerProps={{
        size,
        borderRadiusPercent,
        imgURL,
        color,
        ...containerProps,
        ref: containerRef,
      }}
      shimmerProps={{ borderRadiusPercent, ...shimmerProps, ref: shimmerRef }}
      StyledShimmer={StyledShimmer}
    >
      {children}
      {!imgURL ? (
        <StyledText {...textProps} ref={textRef} size={size} textColor={fontColor}>
          {placeholder}
        </StyledText>
      ) : (
        ''
      )}
    </Skeleton>
  );
};

Avatar.Container = AvatarContainer;
Avatar.Text = AvatarText;

export default Avatar;
