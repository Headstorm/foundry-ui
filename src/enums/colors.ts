enum colors {
  background = '#fff',

  primary = '#DA4200',
  primaryDark = '#BF3A00',
  secondary = '#7349AC',
  secondaryDark = '#512888',
  tertiary = '#64ACC5',
  tertiaryDark = '#338BAA',

  black = '#000',
  grayDark = '#252D34',
  grayMedium = '#51575D',
  grayLight = '#9FA2A5',
  grayXlight = '#D5D5D5',

  success = '#698F00',
  destructive = '#d00',
  shadow = '#14161F',

  grayDark25 = 'rgba(211, 214, 215, 0.25)',
  grayDark50 = 'rgba(211, 214, 215, 0.5)',
  grayDark75 = 'rgba(211, 214, 215, 0.75)',

  transparent = 'transparent',
}

export default colors;
