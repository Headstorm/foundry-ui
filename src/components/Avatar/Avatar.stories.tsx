import * as React from 'react';
import { Story, Meta } from '@storybook/react';
import styled from 'styled-components';
import { withFoundryContext } from '../../../.storybook/decorators';
import colors from '../../enums/colors';
import Avatar, { AvatarProps } from './Avatar';

export const DefaultAvatar: Story<AvatarProps> = (args: AvatarProps) => <Avatar {...args} />;
DefaultAvatar.args = {
  placeholder: 'AA',
  children: '',
  size: 10,
  imgURL: 'https://source.unsplash.com/collection/19271953',
  borderRadiusPercent: 50,
  color: colors.grayXlight,
  isLoading: false,
};

const LabelContainer = styled.div`
  position: absolute;
  bottom: 0;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  color: white;
  padding: 8px 0;
  text-align: center;
`;

export const LabelAvatar: Story<AvatarProps & { children: string }> = ({
  children,
  ...args
}: AvatarProps) => (
  <Avatar {...args}>
    <LabelContainer>{children}</LabelContainer>
  </Avatar>
);

LabelAvatar.args = {
  placeholder: 'JS',
  children: 'John Smith',
  size: 10,
  imgURL: 'https://source.unsplash.com/collection/19271953',
  borderRadiusPercent: 10,
  color: colors.grayXlight,
  isLoading: false,
};

export default {
  title: 'Avatar',
  component: Avatar,
  decorators: [withFoundryContext],
  argTypes: {
    borderRadiusPercent: { control: { type: 'range', min: 0, max: 50, step: 1 } },
  },
} as Meta;
