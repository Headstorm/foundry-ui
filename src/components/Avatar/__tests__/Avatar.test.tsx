import React from 'react';
import Avatar from '../Avatar';
import { render, waitFor, configure } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import colors from '../../../enums/colors';

expect.extend(toHaveNoViolations);
configure({ testIdAttribute: 'data-test-id' });
const testId = 'foundry-avatar';

describe('Avatar', () => {
  it('matches snapshot', async () => {
    const { container } = render(<Avatar color={colors.grayXlight} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
  it('matches snapshot loading state', async () => {
    const { container } = render(<Avatar isLoading color={colors.grayXlight} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar displays correct initials when there is no image', async () => {
    const { container } = render(<Avatar placeholder="SA" color={colors.grayXlight} />);

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar displays correct image when their is an image', async () => {
    const { container } = render(
      <Avatar imgURL="https://tinyurl.com/49dba3d4" color={colors.grayXlight} />,
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar size is correct', async () => {
    const { container } = render(<Avatar size={21} color={colors.grayXlight} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar shape is correct', async () => {
    const { container } = render(<Avatar borderRadiusPercent={30} color={colors.grayXlight} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  describe('Accessibility Tests', () => {
    it('Should pass accessibility test with default props', async () => {
      const component = <Avatar color={colors.grayXlight}></Avatar>;
      const { container } = render(component);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });

  describe('Ref Tests', () => {
    it('avatarText.current should exist', async () => {
      const ref = React.createRef<HTMLSpanElement>();
      const { getByTestId } = render(<Avatar textRef={ref} color={colors.grayXlight} />);
      await waitFor(() => {
        expect(ref.current instanceof HTMLSpanElement).toBeTruthy();
      });
    });
  });
});
