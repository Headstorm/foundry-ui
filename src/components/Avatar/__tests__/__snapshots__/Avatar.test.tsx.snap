// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar avatar displays correct image when their is an image 1`] = `
.c2 {
  opacity: 0;
  background: linear-gradient( 90deg, rgba(162,162,162,0.75), rgba(162,162,162,0.25), rgba(162,162,162,0.75) ) repeat;
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-size: 100vw 100vh;
  background-attachment: fixed;
  border-radius: 0.25rem;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  display: block;
  position: relative;
}

.c0 > * {
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
}

.c1 {
  border-radius: 50%;
  overflow: hidden;
  width: 1.5em;
  height: 1.5em;
  background-image: url(https://tinyurl.com/49dba3d4);
  background-size: cover;
}

.c3 {
  border-radius: 50%;
}

<div>
  <div
    class=" c0 c1"
    color="#D5D5D5"
    size="1.5"
  >
    <div
      class=" c2 c3"
      color="#a2a2a2"
    />
  </div>
</div>
`;

exports[`Avatar avatar displays correct initials when there is no image 1`] = `
.c3 {
  opacity: 0;
  background: linear-gradient( 90deg, rgba(162,162,162,0.75), rgba(162,162,162,0.25), rgba(162,162,162,0.75) ) repeat;
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-size: 100vw 100vh;
  background-attachment: fixed;
  border-radius: 0.25rem;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  display: block;
  position: relative;
}

.c0 > * {
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 50%;
  padding: 1em;
  width: 1.5em;
  height: 1.5em;
  background-color: #D5D5D5;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  color: #51575D;
  font-size: 0.5em;
  font-weight: 600;
}

.c4 {
  border-radius: 50%;
}

<div>
  <div
    class=" c0 c1"
    color="#D5D5D5"
    size="1.5"
  >
    <span
      class=" c2"
      size="1.5"
    >
      SA
    </span>
    <div
      class=" c3 c4"
      color="#a2a2a2"
    />
  </div>
</div>
`;

exports[`Avatar avatar shape is correct 1`] = `
.c3 {
  opacity: 0;
  background: linear-gradient( 90deg, rgba(162,162,162,0.75), rgba(162,162,162,0.25), rgba(162,162,162,0.75) ) repeat;
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-size: 100vw 100vh;
  background-attachment: fixed;
  border-radius: 0.25rem;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  display: block;
  position: relative;
}

.c0 > * {
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 30%;
  padding: 1em;
  width: 1.5em;
  height: 1.5em;
  background-color: #D5D5D5;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  color: #51575D;
  font-size: 0.5em;
  font-weight: 600;
}

.c4 {
  border-radius: 30%;
}

<div>
  <div
    class=" c0 c1"
    color="#D5D5D5"
    size="1.5"
  >
    <span
      class=" c2"
      size="1.5"
    />
    <div
      class=" c3 c4"
      color="#a2a2a2"
    />
  </div>
</div>
`;

exports[`Avatar avatar size is correct 1`] = `
.c3 {
  opacity: 0;
  background: linear-gradient( 90deg, rgba(162,162,162,0.75), rgba(162,162,162,0.25), rgba(162,162,162,0.75) ) repeat;
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-size: 100vw 100vh;
  background-attachment: fixed;
  border-radius: 0.25rem;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  display: block;
  position: relative;
}

.c0 > * {
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 50%;
  padding: 1em;
  width: 21em;
  height: 21em;
  background-color: #D5D5D5;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  color: #51575D;
  font-size: 7em;
  font-weight: 600;
}

.c4 {
  border-radius: 50%;
}

<div>
  <div
    class=" c0 c1"
    color="#D5D5D5"
    size="21"
  >
    <span
      class=" c2"
      size="21"
    />
    <div
      class=" c3 c4"
      color="#a2a2a2"
    />
  </div>
</div>
`;

exports[`Avatar matches snapshot 1`] = `
.c3 {
  opacity: 0;
  background: linear-gradient( 90deg, rgba(162,162,162,0.75), rgba(162,162,162,0.25), rgba(162,162,162,0.75) ) repeat;
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-size: 100vw 100vh;
  background-attachment: fixed;
  border-radius: 0.25rem;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  display: block;
  position: relative;
}

.c0 > * {
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 50%;
  padding: 1em;
  width: 1.5em;
  height: 1.5em;
  background-color: #D5D5D5;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  color: #51575D;
  font-size: 0.5em;
  font-weight: 600;
}

.c4 {
  border-radius: 50%;
}

<div>
  <div
    class=" c0 c1"
    color="#D5D5D5"
    size="1.5"
  >
    <span
      class=" c2"
      size="1.5"
    />
    <div
      class=" c3 c4"
      color="#a2a2a2"
    />
  </div>
</div>
`;

exports[`Avatar matches snapshot loading state 1`] = `
.c4 {
  opacity: 1;
  background: linear-gradient( 90deg, rgba(162,162,162,0.75), rgba(162,162,162,0.25), rgba(162,162,162,0.75) ) repeat;
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-size: 100vw 100vh;
  background-attachment: fixed;
  border-radius: 0.25rem;
  -webkit-animation: none;
  animation: none;
}

.c0 {
  display: block;
  position: relative;
  color: transparent !important;
}

.c0 > * {
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
}

.c0 > *:not(.c3) {
  opacity: 0 !important;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 50%;
  padding: 1em;
  width: 1.5em;
  height: 1.5em;
  background-color: #D5D5D5;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  color: #51575D;
  font-size: 0.5em;
  font-weight: 600;
}

.c5 {
  border-radius: 50%;
}

<div>
  <div
    class=" c0 c1"
    color="#D5D5D5"
    size="1.5"
  >
    <span
      class=" c2"
      size="1.5"
    />
    <div
      class=" c3 c4 c5"
      color="#a2a2a2"
    />
  </div>
</div>
`;
